src/mame/video/kaneko_grap2.cpp

```diff
@@ -14,43 +14,61 @@
 #include "emu.h"
 #include "kaneko_grap2.h"
 
+ADDRESS_MAP_START(kaneko_grap2_device::grap2_map)
+	AM_RANGE(0x000000, 0x0003ff) AM_READWRITE(unk1_r, unk1_w)
+	AM_RANGE(0x000400, 0x000401) AM_WRITE(framebuffer1_scrollx_w)
+	AM_RANGE(0x000800, 0x000bff) AM_READWRITE(unk2_r, unk2_w)
+	AM_RANGE(0x000c00, 0x000c01) AM_WRITE(framebuffer1_scrolly_w)
+	AM_RANGE(0x000c02, 0x000c03) AM_WRITE(framebuffer1_enable_w)
+	AM_RANGE(0x000c06, 0x000c07) AM_WRITE(framebuffer1_bgcol_w)
+	AM_RANGE(0x000c10, 0x000c11) AM_READWRITE(framebuffer1_fbbright1_r, framebuffer1_fbbright1_w )
+	AM_RANGE(0x000c12, 0x000c13) AM_READWRITE(framebuffer1_fbbright2_r, framebuffer1_fbbright2_w )
+	AM_RANGE(0x000c18, 0x000c1b) AM_WRITE(regs1_address_w)
+	AM_RANGE(0x000c1c, 0x000c1d) AM_WRITE(regs2_w)
+	AM_RANGE(0x000c1e, 0x000c1f) AM_WRITE(regs1_go_w)
+	AM_RANGE(0x000c00, 0x000c1f) AM_READ(regs1_r)
+	AM_RANGE(0x080000, 0x0801ff) AM_READWRITE( pal_r, framebuffer1_palette_w )
+	AM_RANGE(0x100000, 0x17ffff) AM_READWRITE( framebuffer_r, framebuffer_w )
+ADDRESS_MAP_END
+
 DEFINE_DEVICE_TYPE(KANEKO_GRAP2, kaneko_grap2_device, "kaneko_grap2", "Kaneko GRAP2")
 
 kaneko_grap2_device::kaneko_grap2_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
 	: device_t(mconfig, KANEKO_GRAP2, tag, owner, clock)
-	, m_palette(*this, finder_base::DUMMY_TAG)
+	, device_rom_interface(mconfig, *this, 32) // TODO : Unknown Address Bits
+	, m_palette(*this, "palette")
 {
-	m_chipnum = 0;
 }
 
+MACHINE_CONFIG_START(kaneko_grap2_device::device_add_mconfig)
+	MCFG_PALETTE_ADD("palette", 0x101)
+	MCFG_PALETTE_FORMAT(xGGGGGRRRRRBBBBB)
+MACHINE_CONFIG_END
 
 void kaneko_grap2_device::device_start()
 {
 	m_framebuffer = make_unique_clear<uint16_t[]>(0x80000/2);
-	m_framebuffer_palette = make_unique_clear<uint16_t[]>(0x200/2);
+	m_framebuffer_palette = make_unique_clear<uint16_t[]>(0x101);
```

Comment:
Why does it need the extra one compared to the previous code?

Reply:
Internal palette 0x100 + Background color 1, It's Gals Panic 3 framebuffer pixel draw routine related.